Fix compilation on C++ project

Rem01Gaming opened the pull request.

Rem01Gaming committed:
Fix compilation on C++ project

In file included from jni/Main.cpp:26:
jni/external/x-watcher/include/x-watcher/x-watcher.h:448:2: error: assigning to 'struct _ArrayHeader *' from incompatible type 'void * _Nullable'
  448 |         arr_init(watcher->directories);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:15:21: note: expanded from macro 'arr_init'
   15 | #define arr_init(a) arr_init_n((a), ARRAY_INITIAL_SIZE)
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:19:13: note: expanded from macro 'arr_init_n'
   19 |    header = malloc(sizeof(*header) + (sizeof(*(a)) * (n))); \
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from jni/Main.cpp:26:
jni/external/x-watcher/include/x-watcher/x-watcher.h:448:2: error: assigning to 'struct directory *' from incompatible type 'void *'
  448 |         arr_init(watcher->directories);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:15:21: note: expanded from macro 'arr_init'
   15 | #define arr_init(a) arr_init_n((a), ARRAY_INITIAL_SIZE)
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:22:10: note: expanded from macro 'arr_init_n'
   22 |    (a) = (void*)(header + 1); \
      |          ^~~~~~~~~~~~~~~~~~~
In file included from jni/Main.cpp:26:
jni/external/x-watcher/include/x-watcher/x-watcher.h:517:3: error: assigning to 'struct _ArrayHeader *' from incompatible type 'void * _Nullable'
  517 |                 arr_init(new_dir.files);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:15:21: note: expanded from macro 'arr_init'
   15 | #define arr_init(a) arr_init_n((a), ARRAY_INITIAL_SIZE)
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:19:13: note: expanded from macro 'arr_init_n'
   19 |    header = malloc(sizeof(*header) + (sizeof(*(a)) * (n))); \
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from jni/Main.cpp:26:
jni/external/x-watcher/include/x-watcher/x-watcher.h:517:3: error: assigning to 'struct file *' from incompatible type 'void *'
  517 |                 arr_init(new_dir.files);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:15:21: note: expanded from macro 'arr_init'
   15 | #define arr_init(a) arr_init_n((a), ARRAY_INITIAL_SIZE)
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:22:10: note: expanded from macro 'arr_init_n'
   22 |    (a) = (void*)(header + 1); \
      |          ^~~~~~~~~~~~~~~~~~~
In file included from jni/Main.cpp:26:
jni/external/x-watcher/include/x-watcher/x-watcher.h:520:3: error: assigning to 'struct directory *' from incompatible type 'void *'
  520 |                 arr_add(watcher->directories, new_dir);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:57:7: note: expanded from macro 'arr_add'
   57 |       arr_reserve((a), header->capacity << 1); \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:35:10: note: expanded from macro 'arr_reserve'
   35 |    (a) = (void*)((struct _ArrayHeader*)realloc( \
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   36 |       header, sizeof(*header) + (sizeof(*(a)) * (n))) + 1); \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from jni/Main.cpp:26:
jni/external/x-watcher/include/x-watcher/x-watcher.h:547:2: error: assigning to 'struct file *' from incompatible type 'void *'
  547 |         arr_add(dir->files, new_file);
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:57:7: note: expanded from macro 'arr_add'
   57 |       arr_reserve((a), header->capacity << 1); \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:35:10: note: expanded from macro 'arr_reserve'
   35 |    (a) = (void*)((struct _ArrayHeader*)realloc( \
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   36 |       header, sizeof(*header) + (sizeof(*(a)) * (n))) + 1); \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from jni/Main.cpp:26:
jni/external/x-watcher/include/x-watcher/x-watcher.h:700:3: error: assigning to 'struct _ArrayHeader *' from incompatible type 'void * _Nullable'
  700 |                 arr_init(dir.files);
      |                 ^~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:15:21: note: expanded from macro 'arr_init'
   15 | #define arr_init(a) arr_init_n((a), ARRAY_INITIAL_SIZE)
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:19:13: note: expanded from macro 'arr_init_n'
   19 |    header = malloc(sizeof(*header) + (sizeof(*(a)) * (n))); \
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from jni/Main.cpp:26:
jni/external/x-watcher/include/x-watcher/x-watcher.h:700:3: error: assigning to 'struct file *' from incompatible type 'void *'
  700 |                 arr_init(dir.files);
      |                 ^~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:15:21: note: expanded from macro 'arr_init'
   15 | #define arr_init(a) arr_init_n((a), ARRAY_INITIAL_SIZE)
      |                     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:22:10: note: expanded from macro 'arr_init_n'
   22 |    (a) = (void*)(header + 1); \
      |          ^~~~~~~~~~~~~~~~~~~
In file included from jni/Main.cpp:26:
jni/external/x-watcher/include/x-watcher/x-watcher.h:805:3: error: assigning to 'struct directory *' from incompatible type 'void *'
  805 |                 arr_add(watcher->directories, dir);
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:57:7: note: expanded from macro 'arr_add'
   57 |       arr_reserve((a), header->capacity << 1); \
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
jni/external/x-watcher/include/x-watcher/array.h:35:10: note: expanded from macro 'arr_reserve'
   35 |    (a) = (void*)((struct _ArrayHeader*)realloc( \
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   36 |       header, sizeof(*header) + (sizeof(*(a)) * (n))) + 1); \
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
9 errors generated.
make: *** [C:/Users/LENOVO/AppData/Local/Android/Sdk/ndk/27.0.12077973/build/../build/core/build-binary.mk:422: obj/local/arm64-v8a/objs/encored/Main.o] Error 1

Signed-off-by: Rem01Gaming <Rem01_Gaming@proton.me>

Rem01Gaming committed:
Fix duplicate symbol for `file` and `directory` struct

Signed-off-by: Rem01Gaming <Rem01_Gaming@proton.me>

nikp123 commented on include/x-watcher/array.h:
```diff
@@ -16,10 +16,10 @@ struct _ArrayHeader {
    header->count    = 0; \
    header->capacity = (n); \
-   (a) = (void*)(header + 1); \
+   (a) = (__typeof__(a))((void*)(header + 1)); \
```
``__typeof__`` seems fairly new (C23). Do we not have alternatives compatible with older versions of C. The idea is to keep this library backwards compatible for at least everything made in the past 10ish years.

nikp123 replied:
See: https://learn.microsoft.com/en-us/cpp/c-language/typeof-c?view=msvc-170

However if gcc/clang already support this for a quite bit longer, then I don't really care about MSVC.

nikp123 replied:
Yea, this is a non-issue. Both gcc and clang support this going way back.

nikp123 commented on include/x-watcher/array.h:
```diff
@@ -63,15 +64,13 @@ struct _ArrayHeader {
    for(size_t i = 0; i < arr_count((a)); i++) { \
       if((a)[i] == val) { \
-         *idx = i; \
+         *(idx) = (int)i; \
```
This can be an unsafe cast if we have a array that has more elements than 32 bits can represent. Ideally we'd use ``size_t`` here, I'd think.

nikp123 replied:
Or just ``__typeof__(idx)`` if that ends up working out.

nikp123 replied:
Yeah, just change it to ``__typeof__(idx)`` so we don't have unsafe casts here. Or at least, make the library user responsible for any safety issues they bring.

Rem01Gaming replied:
Fixed in db8df31ff76267fb9b7853973acfaf80e891f168

nikp123 commented:
The duplicate symbol fix is fine (as it seems to not break anything in prod rn), however the array.h changes need some more considerations.

nikp123 reviewed (commented):
i forgor how to use the review ui again

Rem01Gaming committed:
`arr_find` macro index type safety

Previously used int which could truncate large array indices.
Now uses __typeof__(*(idx)) to match the user's pointer type,
shifting responsibility for overflow to the caller.

Signed-off-by: Rem01Gaming <Rem01_Gaming@proton.me>

nikp123 commented:
Anyway looks good now. Thanks for the fixes.

Rem01Gaming commented:
> Anyway looks good now. Thanks for the fixes.

You're welcome 😊
